pep-0683.rst

```diff
@@ -0,0 +1,342 @@
+  be incref'ed enough to reach ``_Py_IMMORTAL_REFCNT`` and then
+  be treated as immortal
+
+??? immortal != immutable
+
+
+Concerns
+========
+
+(This topic has been discussed previously on python-dev. [python-dev_])
+
+Concerns have centered around the performance penalty.
+Mitigation is discussed above.
+
+
+Specification
+=============
+
+The approach involves these fundamental changes:
+
+* add ``_Py_IMMORTAL_REFCNT`` (the magic value) to the internal C-API
+* update ``Py_INCREF()`` and ``Py_DECREF()`` to no-op for objects with
+  the magic refcount (or its most significant bit)
+* do the same for any other API that modifies the refcount
+* ensure that all immortal objects are cleaned up during
+  runtime finalization
+
+Then setting any object's refcount to ``_Py_IMMORTAL_REFCNT``
+makes it immortal.
+
+To be clear, we will likely use the most-significant bit of
+``_Py_IMMORTAL_REFCNT`` to tell if an object is immortal, rather
+than comparing with ``_Py_IMMORTAL_REFCNT`` directly.
+
+(There are other minor, internal changes which are not described here.)
+
+This is not meant to be a public feature but rather an internal one.
+So the proposal does *not* including adding any new public C-API,
+nor any Python API.  However, this does not prevent us from
+adding (publicly accessible) private API to do things
+like immortalize an object or tell if one
+is immortal.
+
+Affected API
+------------
+
+API that will now ignore immortal objects:
+
+* (public) ``Py_INCREF()``
+* (public) ``Py_DECREF()``
+* (public) ``Py_SET_REFCNT()``
+* (private) ``_Py_NewReference()``
+
+API that exposes refcounts (unchanged but may now return large values):
+
+* (public) ``Py_REFCNT()``
+* (public) ``sys.getrefcount()``
+* (public) ``sys.gettotalrefcount()``
+
+Immortal Global Objects
+-----------------------
+
+The following objects will be made immortal:
+
+* singletons (``None``, ``True``, ``False``, ``Ellipsis``, ``NotImplemented``)
+* all static types (e.g. ``PyLong_Type``, ``PyExc_Exception``)
+* all static objects in ``_PyRuntimeState.global_objects`` (e.g. identifiers,
+  small ints)
+
+There will likely be others we have not enumerated here.
+
+Documentation
+-------------
+
+The feature itself is internal and will not be added to the documentation.
+
+We *may* add a note about immortal objects to the following,
+to help reduce any surprise users may have with the change:
+
+* ``Py_SET_REFCNT()`` (a no-op for immortal objects)
+* ``Py_REFCNT()`` (value may be surprisingly large)
+* ``sys.getrefcount()`` (value may be surprisingly large)
+
+Other API that might benefit from such notes are currently undocumented.
+
+We wouldn't add a note anywhere else (including for ``Py_INCREF()`` and
+``Py_DECREF()``) since the feature is otherwise transparent to users.
+
+
+How to Teach This
+=================
+
+This is not a user-facing change.
+
+
+Reference Implementation
+========================
+
+A PR has been posted. [elizondo]_
```

Comment:
Could you change to something like "The reference implementation is contained in CPython PR #spam` please (and make it an inline link).

Reply:
Why an inline link?

Reply:
Detail is in #2130 et al, the precis is that footnotes add an extra step of indirection. 

To keep all links in one place, you could use `target_` (no square brackets) and change 

```
.. [target]
   https://target.example/spam/spam/ham/eggs/and/spam
```

to

```
.. _target
   https://target.example/spam/spam/ham/eggs/and/spam
```

A

Reply:
Sorry, I meant is there any problem with spelling out the URL directly?